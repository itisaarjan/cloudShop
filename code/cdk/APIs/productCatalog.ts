import * as api from '@aws-sdk/client-api-gateway';

const apiClient = new api.APIGatewayClient({ region: 'us-east-1' });

interface APIprop {
  restAPId: string;
  parentId: string;
  resourceId: string;
}

interface ProductCatalogAPIProps {
  name: string;
  description: string;
  version: string;
}

// Initialize empty API object
let API: Partial<APIprop> = {};

async function createProductCatalogAPI(props: ProductCatalogAPIProps) {
  const command = new api.CreateRestApiCommand({
    name: props.name,
    description: props.description,
    version: props.version,
    binaryMediaTypes: ['application/json']
  });

  try {
    const response = await apiClient.send(command);
    console.log('API created successfully:', response);

    if (!response.id || !response.rootResourceId) {
      throw new Error('API creation response is missing id or rootResourceId');
    }

    API.restAPId = response.id;
    API.parentId = response.rootResourceId;

    return {
      apiId: response.id,
      rootResourceId: response.rootResourceId
    };
  } catch (error: any) {
    console.error('Error creating API:', error);
    throw new Error(`Failed to create API: ${error.message || error}`);
  }
}

async function createProductsResource(apiId: string, rootResourceId: string) {
  const command = new api.CreateResourceCommand({
    restApiId: apiId,
    parentId: rootResourceId,
    pathPart: 'products'
  });

  try {
    const response = await apiClient.send(command);
    console.log('/products resource created:', response);
    API.resourceId = response.id!;
    return response.id!;
  } catch (error: any) {
    console.error('Error creating /products resource:', error);
    throw new Error(`Failed to create /products: ${error.message || error}`);
  }
}

async function createProductIdResource(apiId: string, parentId: string) {
  const command = new api.CreateResourceCommand({
    restApiId: apiId,
    parentId: parentId,
    pathPart: '{productId}'
  });

  try {
    const response = await apiClient.send(command);
    console.log('/products/{productId} resource created:', response);
    return response.id!;
  } catch (error: any) {
    console.error('Error creating /products/{productId} resource:', error);
    throw new Error(`Failed to create /products/{productId}: ${error.message || error}`);
  }
}

// Main function to create everything
async function setupProductCatalogAPI() {
  try {
    const { apiId, rootResourceId } = await createProductCatalogAPI({
      name: 'ProductCatalogAPI',
      description: 'API for managing product catalog',
      version: '1.0'
    });

    const productsResourceId = await createProductsResource(apiId, rootResourceId);
    const productIdResourceId = await createProductIdResource(apiId, productsResourceId);

    console.log(`Resources created under API ID: ${apiId}`);
    console.log(`/products resource ID: ${productsResourceId}`);
    console.log(`/products/{productId} resource ID: ${productIdResourceId}`);
  } catch (err) {
    console.error('Failed to setup API:', err);
  }
}

setupProductCatalogAPI();

async function getProductsAPI(){
  const command = new api.PutMethodCommand({
    restApiId: API.restAPId,
    resourceId: API.resourceId,
    httpMethod: 'GET',
    authorizationType: 'NONE',
  })
}

