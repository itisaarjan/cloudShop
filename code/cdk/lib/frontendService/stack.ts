import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as certificateManager from 'aws-cdk-lib/aws-certificatemanager';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins';
import * as targets from 'aws-cdk-lib/aws-route53-targets';
import { S3Bucket } from 'aws-cdk-lib/aws-kinesisfirehose';

export class FrontendStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const bucket = new s3.Bucket(this, 'CloudShopBucket', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    });

    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
      domainName: 'cloudshop.click',
    });       

    const certificate = certificateManager.Certificate.fromCertificateArn(this, 
      "CloudShopExistingCertificate",
      'arn:aws:acm:us-east-1:985539769349:certificate/d4819f64-1bd4-41f0-ab1c-0f7151dec481');

    const distribution = new cloudfront.Distribution(this, 'CloudFrontDistribution', {
      defaultBehavior: {
        origin: origins.S3BucketOrigin.withOriginAccessControl(bucket),
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      domainNames: ['www.cloudshop.click'],
      certificate,
      errorResponses: [
        {
          httpStatus: 403,
          responseHttpStatus: 200,
          responsePagePath: '/index.html',
          ttl: cdk.Duration.seconds(3),
        },
        {
          httpStatus: 404,
          responseHttpStatus: 200,
          responsePagePath: '/index.html',
          ttl: cdk.Duration.seconds(3),
        },
      ],
    });
  }
}
