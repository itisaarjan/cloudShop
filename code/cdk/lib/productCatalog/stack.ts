import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambda from "aws-cdk-lib/aws-lambda-nodejs";
import * as apigateway from "aws-cdk-lib/aws-apigateway";
import * as integrations from "aws-cdk-lib/aws-apigateway";
import path from 'path';

export class ProductCatalogStack extends cdk.Stack {
 constructor(scope: Construct, id: string, props?: cdk.StackProps) {
   super(scope, id, props);


   const table = new dynamodb.Table(this,"productCatalogTable",{
     tableName: 'productCatalogTable',
     partitionKey: {
       name: 'id',
       type: dynamodb.AttributeType.STRING
     },
     sortKey:{
       name:'name',
       type: dynamodb.AttributeType.STRING
     },
     pointInTimeRecoverySpecification:{
       pointInTimeRecoveryEnabled: true
     },
     billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
     replicationRegions: ["us-west-1","us-west-2"],
    
   })

   const lambdaGetProduct = new lambda.NodejsFunction(this, 'lambdaGetProduct', {
     functionName:"lambdaGetProduct",
     runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
     handler: 'handler',
     entry: path.join(__dirname,"..",'..',"services",'productCatalogServices',"lambda_get_productddb.ts"),
     environment:{
       table_name: table.tableName
     }
   });

   const lambdaGetAllProducts = new lambda.NodejsFunction(this, 'lambdaGetAllProducts',{
    functionName: "lambdaGetAllProducts",
    runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
    handler: 'handler',
    entry: path.join(__dirname,"..","..","services","productCatalogServices",'lambda_get_all_productddb.ts'),
    environment :{
      table_name: table.tableName
    }
   })

   const lambdaPostProduct = new lambda.NodejsFunction(this, 'lambdaPostProduct',{
    functionName: 'lambdaPostProudct',
    runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
    handler: 'handler',
    entry: path.join(__dirname, '..','..', 'services','productCatalogServices', 'lambda_post_productddb.ts'),
    environment: {
      table_name: table.tableName
    }
   })

   const lambdaDeleteProduct = new lambda.NodejsFunction(this, 'lambdaDeleteProduct', {
     functionName: 'lambdaDeleteProduct',
     runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
     handler: 'handler',
     entry: path.join(__dirname, '..','..', 'services','productCatalogServices', 'lambda_delete_productddb.ts'),
     environment: {
       table_name: table.tableName
     }
   });

   const lambdaUpdateProduct = new lambda.NodejsFunction(this, 'lambdaUpdateProduct', {
     functionName: 'lambdaUpdateProduct',
     runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
     handler: 'handler',
     entry: path.join(__dirname, '..','..', 'services','productCatalogServices', 'lambda_update_productddb.ts'),
     environment: {
       table_name: table.tableName
     }
   });

   table.grantReadData(lambdaGetProduct);
   table.grantWriteData(lambdaPostProduct);
   table.grantReadWriteData(lambdaDeleteProduct);
   table.grantReadWriteData(lambdaUpdateProduct);

   const api = new apigateway.RestApi(this, "Products");
   const v1 = api.root.addResource("v1");
   const products = v1.addResource("products");
   const productId = products.addResource("{id}");
   const productName = productId.addResource("{name}");

   productName.addMethod('GET', new integrations.LambdaIntegration(lambdaGetProduct));
   products.addMethod('Get', new integrations.LambdaIntegration(lambdaGetAllProducts));
   products.addMethod('POST', new integrations.LambdaIntegration(lambdaPostProduct));
   products.addMethod('PUT', new integrations.LambdaIntegration(lambdaUpdateProduct));
   products.addMethod('DELETE', new integrations.LambdaIntegration(lambdaDeleteProduct));
 }
}



