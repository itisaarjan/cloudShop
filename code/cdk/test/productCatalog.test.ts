import * as cdk from 'aws-cdk-lib';


import { Template } from 'aws-cdk-lib/assertions';
import { ProductCatalogStack } from '../lib/productCatalog/stack';

describe('Snapshot testing to detect sudden infra changes',()=>{
    const testApp = new cdk.App({
        outdir: 'cdk.out',
    });

    const ProductCatalogTestStack = new ProductCatalogStack(testApp,'ProductCatalogTest');

    const ProductCatalogStacktemplate = Template.fromStack(ProductCatalogTestStack);
    
    test('Snapshot test',()=>{
        expect(ProductCatalogStacktemplate.toJSON()).toMatchSnapshot();
    })
})