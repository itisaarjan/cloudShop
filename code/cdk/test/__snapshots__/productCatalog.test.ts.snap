// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing to detect sudden infra changes Snapshot test 1`] = `
{
  "Conditions": {
    "productCatalogTableStackRegionNotEqualsuswest1F82DADEE": {
      "Fn::Not": [
        {
          "Fn::Equals": [
            "us-west-1",
            {
              "Ref": "AWS::Region",
            },
          ],
        },
      ],
    },
    "productCatalogTableStackRegionNotEqualsuswest2F5F48711": {
      "Fn::Not": [
        {
          "Fn::Equals": [
            "us-west-2",
            {
              "Ref": "AWS::Region",
            },
          ],
        },
      ],
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "awscdkawsdynamodbReplicaProviderNestedStackawscdkawsdynamodbReplicaProviderNestedStackResource18E3F12D": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Parameters": {
          "referencetoProductCatalogTestproductCatalogTableB1517EA5Ref": {
            "Ref": "productCatalogTable910C8E36",
          },
        },
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region",
              },
              ".",
              {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              {
                "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
              },
              "/0f61ab6344bb42b689ade3486f65ca0838f3f0d9bdf11cf8e8ff9eaa1a5b919f.json",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "lambdaDeleteProduct62D4A3CA": {
      "DependsOn": [
        "lambdaDeleteProductServiceRoleDefaultPolicyEF414FAD",
        "lambdaDeleteProductServiceRoleD08376CB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "86fbd58fc898fa353ef0dc05ad53b2b6ec8772bb69466f9411666f5160f29b33.zip",
        },
        "Environment": {
          "Variables": {
            "table_name": {
              "Ref": "productCatalogTable910C8E36",
            },
          },
        },
        "FunctionName": "lambdaDeleteProduct",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "lambdaDeleteProductServiceRoleD08376CB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaDeleteProductServiceRoleD08376CB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaDeleteProductServiceRoleDefaultPolicyEF414FAD": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productCatalogTable910C8E36",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-2:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaDeleteProductServiceRoleDefaultPolicyEF414FAD",
        "Roles": [
          {
            "Ref": "lambdaDeleteProductServiceRoleD08376CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "lambdaGetProduct4AF93531": {
      "DependsOn": [
        "lambdaGetProductServiceRoleDefaultPolicyC8E1DE32",
        "lambdaGetProductServiceRoleCB3852A4",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2574445a0ad436cf47f2b2987c9e295a9a50281daef6c269c1897dcba4ec846d.zip",
        },
        "Environment": {
          "Variables": {
            "table_name": {
              "Ref": "productCatalogTable910C8E36",
            },
          },
        },
        "FunctionName": "lambdaGetProduct",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "lambdaGetProductServiceRoleCB3852A4",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaGetProductServiceRoleCB3852A4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaGetProductServiceRoleDefaultPolicyC8E1DE32": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productCatalogTable910C8E36",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-2:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaGetProductServiceRoleDefaultPolicyC8E1DE32",
        "Roles": [
          {
            "Ref": "lambdaGetProductServiceRoleCB3852A4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "lambdaPostProductE8872313": {
      "DependsOn": [
        "lambdaPostProductServiceRoleDefaultPolicy2179AD7B",
        "lambdaPostProductServiceRole2CBA2639",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "25efe2589d7b2fa36f9cd3a1096649f828af02b92f5d2e9e8906f2673ee549bb.zip",
        },
        "Environment": {
          "Variables": {
            "table_name": {
              "Ref": "productCatalogTable910C8E36",
            },
          },
        },
        "FunctionName": "lambdaPostProudct",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "lambdaPostProductServiceRole2CBA2639",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaPostProductServiceRole2CBA2639": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaPostProductServiceRoleDefaultPolicy2179AD7B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productCatalogTable910C8E36",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-2:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaPostProductServiceRoleDefaultPolicy2179AD7B",
        "Roles": [
          {
            "Ref": "lambdaPostProductServiceRole2CBA2639",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "lambdaUpdateProductB1BC3561": {
      "DependsOn": [
        "lambdaUpdateProductServiceRoleDefaultPolicy6230322B",
        "lambdaUpdateProductServiceRole8F8F185A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "31415d359614ee5e1a30bdb22a31061ba933c2a0f36a157adf19ea163d26a935.zip",
        },
        "Environment": {
          "Variables": {
            "table_name": {
              "Ref": "productCatalogTable910C8E36",
            },
          },
        },
        "FunctionName": "lambdaUpdateProduct",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "lambdaUpdateProductServiceRole8F8F185A",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaUpdateProductServiceRole8F8F185A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaUpdateProductServiceRoleDefaultPolicy6230322B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productCatalogTable910C8E36",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-2:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaUpdateProductServiceRoleDefaultPolicy6230322B",
        "Roles": [
          {
            "Ref": "lambdaUpdateProductServiceRole8F8F185A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "productCatalogTable910C8E36": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          {
            "AttributeName": "name",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "name",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableName": "productCatalogTable",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "productCatalogTableReplicauswest18430B910": {
      "Condition": "productCatalogTableStackRegionNotEqualsuswest1F82DADEE",
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "productCatalogTableSourceTableAttachedManagedPolicyProductCatalogTestawscdkawsdynamodbReplicaProviderIsCompleteHandlerServiceRoleC26D9E60A592FF1F",
        "productCatalogTableSourceTableAttachedManagedPolicyProductCatalogTestawscdkawsdynamodbReplicaProviderOnEventHandlerServiceRole5CA0DF4E41FACC74",
      ],
      "Properties": {
        "Region": "us-west-1",
        "ServiceToken": {
          "Fn::GetAtt": [
            "awscdkawsdynamodbReplicaProviderNestedStackawscdkawsdynamodbReplicaProviderNestedStackResource18E3F12D",
            "Outputs.ProductCatalogTestawscdkawsdynamodbReplicaProviderframeworkonEvent4F2FF11DArn",
          ],
        },
        "TableName": {
          "Ref": "productCatalogTable910C8E36",
        },
      },
      "Type": "Custom::DynamoDBReplica",
      "UpdateReplacePolicy": "Delete",
    },
    "productCatalogTableReplicauswest23DA2D40D": {
      "Condition": "productCatalogTableStackRegionNotEqualsuswest2F5F48711",
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "productCatalogTableSourceTableAttachedManagedPolicyProductCatalogTestawscdkawsdynamodbReplicaProviderIsCompleteHandlerServiceRoleC26D9E60A592FF1F",
        "productCatalogTableSourceTableAttachedManagedPolicyProductCatalogTestawscdkawsdynamodbReplicaProviderOnEventHandlerServiceRole5CA0DF4E41FACC74",
      ],
      "Metadata": {
        "DynamoDbReplicationDependency": {
          "Fn::If": [
            "productCatalogTableStackRegionNotEqualsuswest1F82DADEE",
            {
              "Ref": "productCatalogTableReplicauswest18430B910",
            },
            {
              "Ref": "AWS::NoValue",
            },
          ],
        },
      },
      "Properties": {
        "Region": "us-west-2",
        "ServiceToken": {
          "Fn::GetAtt": [
            "awscdkawsdynamodbReplicaProviderNestedStackawscdkawsdynamodbReplicaProviderNestedStackResource18E3F12D",
            "Outputs.ProductCatalogTestawscdkawsdynamodbReplicaProviderframeworkonEvent4F2FF11DArn",
          ],
        },
        "TableName": {
          "Ref": "productCatalogTable910C8E36",
        },
      },
      "Type": "Custom::DynamoDBReplica",
      "UpdateReplacePolicy": "Delete",
    },
    "productCatalogTableSourceTableAttachedManagedPolicyProductCatalogTestawscdkawsdynamodbReplicaProviderIsCompleteHandlerServiceRoleC26D9E60A592FF1F": {
      "Properties": {
        "Description": {
          "Fn::Join": [
            "",
            [
              "DynamoDB replication managed policy for table ",
              {
                "Ref": "productCatalogTable910C8E36",
              },
            ],
          ],
        },
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:DescribeTable",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productCatalogTable910C8E36",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "Roles": [
          {
            "Fn::GetAtt": [
              "awscdkawsdynamodbReplicaProviderNestedStackawscdkawsdynamodbReplicaProviderNestedStackResource18E3F12D",
              "Outputs.ProductCatalogTestawscdkawsdynamodbReplicaProviderIsCompleteHandlerServiceRole2520239BRef",
            ],
          },
        ],
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "productCatalogTableSourceTableAttachedManagedPolicyProductCatalogTestawscdkawsdynamodbReplicaProviderOnEventHandlerServiceRole5CA0DF4E41FACC74": {
      "Properties": {
        "Description": {
          "Fn::Join": [
            "",
            [
              "DynamoDB replication managed policy for table ",
              {
                "Ref": "productCatalogTable910C8E36",
              },
            ],
          ],
        },
        "Path": "/",
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:*",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "productCatalogTable910C8E36",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "dynamodb:*",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-1:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":dynamodb:us-west-2:",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      {
                        "Ref": "productCatalogTable910C8E36",
                      },
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "Roles": [
          {
            "Fn::GetAtt": [
              "awscdkawsdynamodbReplicaProviderNestedStackawscdkawsdynamodbReplicaProviderNestedStackResource18E3F12D",
              "Outputs.ProductCatalogTestawscdkawsdynamodbReplicaProviderOnEventHandlerServiceRoleA969C0AARef",
            ],
          },
        ],
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
